mono/metadata/class.c

```diff
@@ -1889,6 +1889,10 @@ mono_class_layout_fields (MonoClass *klass, int base_instance_size, int packing_
 
 		if (!(field->type->attrs & FIELD_ATTRIBUTE_STATIC)) {
 			ftype = mono_type_get_underlying_type (field->type);
+			if (!ftype) {
+				mono_class_set_type_load_failure (klass, "Field '%s' is an enum type with a bad underlying type", field->name);
+				continue;
+			}
```

Comment:
So there are two cases when that function can return NULL:
- the type is incomplete
- the type has failed to load
Neither case should happen since mono_class_layout_fields () should not be called for these types.